### ISSUE TITLE
Barrier to use of memoization is high/complex

### ISSUE THREAD

**Issue by simster7** [argoproj/argo-workflows#5433]
From a user on Slack:

> My issues were mostly volume-related. Going off my memory since it was about a month ago that I was working on this — the first issue was translating the examples from a single workflow to a top-level workflow that references a workflow template with a memoized step. I didn’t initially figure out that you needed to declare the volume in the top-level workflow and you could declare the volume in the workflowtemplate but it wouldn’t do anything.
The next issue that did end up being a blocker for me was that the configuration that worked out-of-the-box with a single memoized step ran into issues running multiple memoized steps in parallel. I ended up having to move on to different work because figuring out memoization wasn’t high enough priority to put in the time to troubleshoot that.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by terrytangyuan** [comment]
Another perspective is that once we start using it, the maintenance part of memoization caches which I think https://github.com/argoproj/argo-workflows/issues/5406 can help resolve.

**Comment by Joibel** [comment]
Closing as stale. This was opened in 2021 as a catch-all for memoization usability, based on a second-hand Slack report, and the concrete problems it mentions have since been resolved:

- Memoization no longer requires the step to have outputs (3.5+), which was
the most common cause of "my memoized step wasn't memoized". See the [memoization docs](https://argo-workflows.readthedocs.io/en/latest/memoization/).
- Running several memoized steps in parallel against the same cache
ConfigMap now retries on conflict instead of failing.
- Cache maintenance is covered by memoization cache GC (`CACHE_GC_PERIOD`,`CACHE_GC_AFTER_NOT_HIT_DURATION`), which came out of #5406.

The "you have to declare the volume in the top-level workflow" part describes [work avoidance](https://argo-workflows.readthedocs.io/en/latest/work-avoidance/), which uses a PVC; memoization's ConfigMap cache needs no volume. The docs now distinguish the two.

If something about memoization is still unclear or badly documented, please open a new issue naming the specific gap, which is much easier to act on than this umbrella issue. Docs PRs are very welcome.
